### ISSUE TITLE
Memory problem with 80 000 point discretisation, and problem with NetCDF, Trilinos, and Peridigm installation

### ISSUE THREAD

**Issue by theodordesousa** [peridigm/peridigm#114]
Hey. I have two issues I would need some guidance regarding. 
 

### Issue with global tangent matrix
On the stationary computer with Ubuntu 18.04 LTS, running Peridigm 1.5.0, I'm trying to increase the number of points in the bar from the wave-in-bar example. I have divided the bar into two blocks, using elastic-plastic hardening correspondence material models in both. The solver is quasi-static and no damage or contact models are being used. 

Modifying the wave-in-bar discretization generation file, I add five point to the depth of the bar, while maintaining the same aspect ratio given for the example. This works, but is really slow to build the global tangent matrix. When I increase the number of points along the depth to 20, the memory fills up and crashes the computer. I'm running 15 GB of RAM, and tried increasing my SWAP to 32 GB, but this was also capped. 

I would expect that 80 000 points shouldn't require 47 GB of memory, so obviously I'm doing something weird, and some guidance regarding this would be appreciated. 

Or the Peridigm installation is defect:
First of all, I didn't perform the install, according to my supervisor it took her PhD student a month to do it. That student is no longer at the department, and since I am doing my master thesis, I am reluctant to try to reinstall it in that case, as I need to be finished by June.
Running the tests, I get a 36 % pass rate. However, it seems to be due to Peridigm not being able to find the test files. I'll append them below. Some returns an error due to not finding files, while some compares and finds "files are the same" due to the test script disregarding all variables missing from one of the files, which is all variables if one of the files are missing.  Feel free to investigate them below. Some are renamed since they had the same name when merging them into my folder. 
[NOX_QS_Newton_AztecOOPrec.log](https://github.com/peridigm/peridigm/files/4208105/NOX_QS_Newton_AztecOOPrec.log)
[Bar1.log](https://github.com/peridigm/peridigm/files/4208106/Bar1.log)
[Compression_2x1x1_2.log](https://github.com/peridigm/peridigm/files/4208107/Compression_2x1x1_2.log)
[BondBreakingInitialVelocity2.log](https://github.com/peridigm/peridigm/files/4208108/BondBreakingInitialVelocity2.log)
[IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_2.log](https://github.com/peridigm/peridigm/files/4208109/IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_2.log)
[CompressionImplicit_2x1x1_2.log](https://github.com/peridigm/peridigm/files/4208110/CompressionImplicit_2x1x1_2.log)
[Contact_Cubes_Interaction_Blocks.log](https://github.com/peridigm/peridigm/files/4208111/Contact_Cubes_Interaction_Blocks.log)
[Compression_QS_3x2x2_Exodus2.log](https://github.com/peridigm/peridigm/files/4208112/Compression_QS_3x2x2_Exodus2.log)
[Compression_NLCGQS_3x2x2.log](https://github.com/peridigm/peridigm/files/4208113/Compression_NLCGQS_3x2x2.log)
[Compression_NLCGQS_3x2x2_2.log](https://github.com/peridigm/peridigm/files/4208114/Compression_NLCGQS_3x2x2_2.log)
[ElasticPlasticCorrespondenceFullyPrescribedTension.log](https://github.com/peridigm/peridigm/files/4208115/ElasticPlasticCorrespondenceFullyPrescribedTension.log)
[ElasticCorrespondenceFullyPrescribedTension.log](https://github.com/peridigm/peridigm/files/4208116/ElasticCorrespondenceFullyPrescribedTension.log)
[Bar11.log](https://github.com/peridigm/peridigm/files/4208117/Bar11.log)
[Bar8.log](https://github.com/peridigm/peridigm/files/4208118/Bar8.log)
[Contact_Perforation.log](https://github.com/peridigm/peridigm/files/4208119/Contact_Perforation.log)
[Multiphysics_QS_3x2x2.log](https://github.com/peridigm/peridigm/files/4208120/Multiphysics_QS_3x2x2.log)
[Bar.log](https://github.com/peridigm/peridigm/files/4208121/Bar.log)
[ViscoplasticNeedlemanFullyPrescribedTension_WithFlaw_2.log](https://github.com/peridigm/peridigm/files/4208122/ViscoplasticNeedlemanFullyPrescribedTension_WithFlaw_2.log)
[PrecrackedPlate2_2_2.log](https://github.com/peridigm/peridigm/files/4208123/PrecrackedPlate2_2_2.log)
[Compression_QS_3x2x2_Exodus.log](https://github.com/peridigm/peridigm/files/4208124/Compression_QS_3x2x2_Exodus.log)
[ep_cube2.log](https://github.com/peridigm/peridigm/files/4208125/ep_cube2.log)
[DeformationGradient_NoDeformation2.log](https://github.com/peridigm/peridigm/files/4208126/DeformationGradient_NoDeformation2.log)
[SimpleImplicitDiffusion.log](https://github.com/peridigm/peridigm/files/4208127/SimpleImplicitDiffusion.log)
[ViscoplasticNeedlemanFullyPrescribedTension_WithFlaw.log](https://github.com/peridigm/peridigm/files/4208128/ViscoplasticNeedlemanFullyPrescribedTension_WithFlaw.log)
[DeformationGradient_NoLoad.log](https://github.com/peridigm/peridigm/files/4208129/DeformationGradient_NoLoad.log)
[NOX_QS_MatrixFree_FullTangentPrec.log](https://github.com/peridigm/peridigm/files/4208130/NOX_QS_MatrixFree_FullTangentPrec.log)
[NOX_QS_MatrixFree_3x3Prec.log](https://github.com/peridigm/peridigm/files/4208131/NOX_QS_MatrixFree_3x3Prec.log)
[Contact_Ring2.log](https://github.com/peridigm/peridigm/files/4208132/Contact_Ring2.log)
[ViscoplasticNeedlemanFullyPrescribedTension_NoFlaw_2.log](https://github.com/peridigm/peridigm/files/4208133/ViscoplasticNeedlemanFullyPrescribedTension_NoFlaw_2.log)
[Body_Force_Implicit.log](https://github.com/peridigm/peridigm/files/4208134/Body_Force_Implicit.log)
[WaveInBar.log](https://github.com/peridigm/peridigm/files/4208135/WaveInBar.log)
[DeformationGradient_SimpleShear.log](https://github.com/peridigm/peridigm/files/4208136/DeformationGradient_SimpleShear.log)
[CentrifugalLoad.log](https://github.com/peridigm/peridigm/files/4208137/CentrifugalLoad.log)
[NOX_QS_Newton_AztecOOPrec2.log](https://github.com/peridigm/peridigm/files/4208138/NOX_QS_Newton_AztecOOPrec2.log)
[NOX_QS_MatrixFree_NoPrec_2.log](https://github.com/peridigm/peridigm/files/4208139/NOX_QS_MatrixFree_NoPrec_2.log)
[DeformationGradient_PlaneStrainCompression.log](https://github.com/peridigm/peridigm/files/4208140/DeformationGradient_PlaneStrainCompression.log)
[NOX_QS_Newton_NoPrec.log](https://github.com/peridigm/peridigm/files/4208141/NOX_QS_Newton_NoPrec.log)
[ElasticCorrespondenceQSTension.log](https://github.com/peridigm/peridigm/files/4208142/ElasticCorrespondenceQSTension.log)
[NOX_QS_Newton_NoPrec_2.log](https://github.com/peridigm/peridigm/files/4208143/NOX_QS_Newton_NoPrec_2.log)
[DiskFilter2.log](https://github.com/peridigm/peridigm/files/4208144/DiskFilter2.log)
[Compression_QS_Explicit_MultiFreqOutput_3x2x2.log](https://github.com/peridigm/peridigm/files/4208145/Compression_QS_Explicit_MultiFreqOutput_3x2x2.log)
[Body_Force_Implicit2.log](https://github.com/peridigm/peridigm/files/4208146/Body_Force_Implicit2.log)
[PrecrackedPlate2_2.log](https://github.com/peridigm/peridigm/files/4208147/PrecrackedPlate2_2.log)
[ThermalExpansionCube.log](https://github.com/peridigm/peridigm/files/4208148/ThermalExpansionCube.log)
[Notched_Plate_Data_Loader2.log](https://github.com/peridigm/peridigm/files/4208149/Notched_Plate_Data_Loader2.log)
[VariableHorizonParsing2.log](https://github.com/peridigm/peridigm/files/4208150/VariableHorizonParsing2.log)
[Compression_3x1x1.log](https://github.com/peridigm/peridigm/files/4208151/Compression_3x1x1.log)
[Compression_3x1x1_3.log](https://github.com/peridigm/peridigm/files/4208152/Compression_3x1x1_3.log)
[DeformationGradient_PlaneStrainCompression_2.log](https://github.com/peridigm/peridigm/files/4208153/DeformationGradient_PlaneStrainCompression_2.log)
[Bar2.log](https://github.com/peridigm/peridigm/files/4208154/Bar2.log)
[MultipleOutputFiles.log](https://github.com/peridigm/peridigm/files/4208155/MultipleOutputFiles.log)
[Compression_QS_3x2x2_TextFile.log](https://github.com/peridigm/peridigm/files/4208156/Compression_QS_3x2x2_TextFile.log)
[MultipleOutputFiles2.log](https://github.com/peridigm/peridigm/files/4208157/MultipleOutputFiles2.log)
[Contact_Friction3.log](https://github.com/peridigm/peridigm/files/4208158/Contact_Friction3.log)
[DeformationGradient_NoDeformation.log](https://github.com/peridigm/peridigm/files/4208159/DeformationGradient_NoDeformation.log)
[Contact_2x1x1.log](https://github.com/peridigm/peridigm/files/4208160/Contact_2x1x1.log)
[Body_Force_QS_NOX2.log](https://github.com/peridigm/peridigm/files/4208161/Body_Force_QS_NOX2.log)
[DeformationGradient_PureShear.log](https://github.com/peridigm/peridigm/files/4208162/DeformationGradient_PureShear.log)
[Body_Force_QS2.log](https://github.com/peridigm/peridigm/files/4208163/Body_Force_QS2.log)
[Contact_Ring.log](https://github.com/peridigm/peridigm/files/4208164/Contact_Ring.log)
[Compression_QS_3x2x2.log](https://github.com/peridigm/peridigm/files/4208165/Compression_QS_3x2x2.log)
[ThermalExpansionBondFailure.log](https://github.com/peridigm/peridigm/files/4208166/ThermalExpansionBondFailure.log)
[CompressionImplicit_2x1x1.log](https://github.com/peridigm/peridigm/files/4208167/CompressionImplicit_2x1x1.log)
[Pals_Simple_Shear.log](https://github.com/peridigm/peridigm/files/4208168/Pals_Simple_Shear.log)
[Compression_QS_Explicit_3x2x2.log](https://github.com/peridigm/peridigm/files/4208169/Compression_QS_Explicit_3x2x2.log)
[Bar9.log](https://github.com/peridigm/peridigm/files/4208170/Bar9.log)
[Interfaces2.log](https://github.com/peridigm/peridigm/files/4208171/Interfaces2.log)
[Contact_Cubes2.log](https://github.com/peridigm/peridigm/files/4208172/Contact_Cubes2.log)
[BondBreakingInitialVelocity3.log](https://github.com/peridigm/peridigm/files/4208173/BondBreakingInitialVelocity3.log)
[Contact_Cubes.log](https://github.com/peridigm/peridigm/files/4208174/Contact_Cubes.log)
[NOX_QS_MatrixFree_3x3Prec2.log](https://github.com/peridigm/peridigm/files/4208175/NOX_QS_MatrixFree_3x3Prec2.log)
[NodalVariableOutput.log](https://github.com/peridigm/peridigm/files/4208176/NodalVariableOutput.log)
[Contact_Friction4.log](https://github.com/peridigm/peridigm/files/4208177/Contact_Friction4.log)
[Compression_QS_CyclicLoading_3x2x2_2.log](https://github.com/peridigm/peridigm/files/4208178/Compression_QS_CyclicLoading_3x2x2_2.log)
[DefaultBlocks2.log](https://github.com/peridigm/peridigm/files/4208179/DefaultBlocks2.log)
[VariableHorizonParsing.log](https://github.com/peridigm/peridigm/files/4208180/VariableHorizonParsing.log)
[DiskFilter.log](https://github.com/peridigm/peridigm/files/4208181/DiskFilter.log)
[IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log](https://github.com/peridigm/peridigm/files/4208182/IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log)
[MultipleHorizons.log](https://github.com/peridigm/peridigm/files/4208183/MultipleHorizons.log)
[BondBreakingInitialVelocity.log](https://github.com/peridigm/peridigm/files/4208184/BondBreakingInitialVelocity.log)
[Body_Force_QS.log](https://github.com/peridigm/peridigm/files/4208185/Body_Force_QS.log)
[BondBreakingInitialVelocity4.log](https://github.com/peridigm/peridigm/files/4208186/BondBreakingInitialVelocity4.log)
[Notched_Plate_Data_Loader.log](https://github.com/peridigm/peridigm/files/4208187/Notched_Plate_Data_Loader.log)
[CompressionImplicitEssentialBC_2x1x1_2.log](https://github.com/peridigm/peridigm/files/4208188/CompressionImplicitEssentialBC_2x1x1_2.log)
[ElasticPlasticCorrespondenceFullyPrescribedTension_2.log](https://github.com/peridigm/peridigm/files/4208189/ElasticPlasticCorrespondenceFullyPrescribedTension_2.log)
[Bar4.log](https://github.com/peridigm/peridigm/files/4208190/Bar4.log)
[AdaptiveQuasiStaticSolver.log](https://github.com/peridigm/peridigm/files/4208191/AdaptiveQuasiStaticSolver.log)
[ThermalExpansionCube2.log](https://github.com/peridigm/peridigm/files/4208192/ThermalExpansionCube2.log)
[Compression_QS_3x2x2_2.log](https://github.com/peridigm/peridigm/files/4208193/Compression_QS_3x2x2_2.log)
[Compression_2x1x1.log](https://github.com/peridigm/peridigm/files/4208194/Compression_2x1x1.log)
[CompressionInitDisp_2x1x1_2.log](https://github.com/peridigm/peridigm/files/4208195/CompressionInitDisp_2x1x1_2.log)
[NOX_QS_MatrixFree_FullTangentPrec2.log](https://github.com/peridigm/peridigm/files/4208196/NOX_QS_MatrixFree_FullTangentPrec2.log)
[Interfaces.log](https://github.com/peridigm/peridigm/files/4208197/Interfaces.log)
[CompressionInitDisp_2x1x1.log](https://github.com/peridigm/peridigm/files/4208198/CompressionInitDisp_2x1x1.log)
[Contact_Friction2.log](https://github.com/peridigm/peridigm/files/4208199/Contact_Friction2.log)
[Body_Force_QS_NOX.log](https://github.com/peridigm/peridigm/files/4208200/Body_Force_QS_NOX.log)
[Compression_QS_CyclicLoading_3x2x2.log](https://github.com/peridigm/peridigm/files/4208201/Compression_QS_CyclicLoading_3x2x2.log)
[BondBreakingInitialVelocity6.log](https://github.com/peridigm/peridigm/files/4208202/BondBreakingInitialVelocity6.log)
[ViscoplasticNeedlemanFullyPrescribedTension_NoFlaw.log](https://github.com/peridigm/peridigm/files/4208203/ViscoplasticNeedlemanFullyPrescribedTension_NoFlaw.log)
[CompressionImplicitEssentialBC_2x1x1.log](https://github.com/peridigm/peridigm/files/4208204/CompressionImplicitEssentialBC_2x1x1.log)
[Body_Force_Explicit2.log](https://github.com/peridigm/peridigm/files/4208205/Body_Force_Explicit2.log)
[Body_Force_Explicit.log](https://github.com/peridigm/peridigm/files/4208206/Body_Force_Explicit.log)
[PrecrackedPlate2.log](https://github.com/peridigm/peridigm/files/4208207/PrecrackedPlate2.log)
[Contact_Friction.log](https://github.com/peridigm/peridigm/files/4208208/Contact_Friction.log)
[CentrifugalLoad2.log](https://github.com/peridigm/peridigm/files/4208209/CentrifugalLoad2.log)
[SimpleImplicitDiffusion2.log](https://github.com/peridigm/peridigm/files/4208210/SimpleImplicitDiffusion2.log)
[Contact_Cubes_Interaction_Blocks2.log](https://github.com/peridigm/peridigm/files/4208211/Contact_Cubes_Interaction_Blocks2.log)
[Contact_Perforation2.log](https://github.com/peridigm/peridigm/files/4208212/Contact_Perforation2.log)
[WaveInBar_MultiBlock_2.log](https://github.com/peridigm/peridigm/files/4208213/WaveInBar_MultiBlock_2.log)
[IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log](https://github.com/peridigm/peridigm/files/4208214/IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log)
[DeformationGradient_SolidRotation2.log](https://github.com/peridigm/peridigm/files/4208215/DeformationGradient_SolidRotation2.log)
[Multiphysics_QS_3x2x2_2.log](https://github.com/peridigm/peridigm/files/4208216/Multiphysics_QS_3x2x2_2.log)
[NOX_QS_MatrixFree_NoPrec.log](https://github.com/peridigm/peridigm/files/4208217/NOX_QS_MatrixFree_NoPrec.log)
[DeformationGradient_SolidRotation.log](https://github.com/peridigm/peridigm/files/4208218/DeformationGradient_SolidRotation.log)
[Bar6.log](https://github.com/peridigm/peridigm/files/4208219/Bar6.log)
[Compression_3x1x1_InfluenceFunction2.log](https://github.com/peridigm/peridigm/files/4208220/Compression_3x1x1_InfluenceFunction2.log)
[Contact_2x1x1_2.log](https://github.com/peridigm/peridigm/files/4208221/Contact_2x1x1_2.log)
[Compression_3x1x1_InfluenceFunction.log](https://github.com/peridigm/peridigm/files/4208222/Compression_3x1x1_InfluenceFunction.log)
[ep_cube.log](https://github.com/peridigm/peridigm/files/4208223/ep_cube.log)
[Bar5.log](https://github.com/peridigm/peridigm/files/4208224/Bar5.log)
[Bar10.log](https://github.com/peridigm/peridigm/files/4208225/Bar10.log)
[WaveInBar_MultiBlock.log](https://github.com/peridigm/peridigm/files/4208226/WaveInBar_MultiBlock.log)
[Compression_3x1x1_2.log](https://github.com/peridigm/peridigm/files/4208227/Compression_3x1x1_2.log)
[DeformationGradient_NoLoad2.log](https://github.com/peridigm/peridigm/files/4208228/DeformationGradient_NoLoad2.log)
[ElasticCorrespondenceFullyPrescribedTension_2.log](https://github.com/peridigm/peridigm/files/4208229/ElasticCorrespondenceFullyPrescribedTension_2.log)
[Bar7.log](https://github.com/peridigm/peridigm/files/4208230/Bar7.log)
[WaveInBar2.log](https://github.com/peridigm/peridigm/files/4208231/WaveInBar2.log)
[BondBreakingInitialVelocity5.log](https://github.com/peridigm/peridigm/files/4208232/BondBreakingInitialVelocity5.log)
[IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_2.log](https://github.com/peridigm/peridigm/files/4208233/IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_2.log)
[PrecrackedPlate.log](https://github.com/peridigm/peridigm/files/4208234/PrecrackedPlate.log)
[DefaultBlocks.log](https://github.com/peridigm/peridigm/files/4208235/DefaultBlocks.log)
[Compression_QS_Explicit_MultiFreqOutput_3x2x2_2.log](https://github.com/peridigm/peridigm/files/4208236/Compression_QS_Explicit_MultiFreqOutput_3x2x2_2.log)
[Bar3.log](https://github.com/peridigm/peridigm/files/4208237/Bar3.log)
[MultipleHorizons2.log](https://github.com/peridigm/peridigm/files/4208238/MultipleHorizons2.log)

### Issue with peridigm installation on my laptop

Parallel to this I have tried to reinstall Peridigm on my laptop with Ubuntu 18.04 LTS, a 4 core Intel I7 processor. Here, I run into issues starting with NetCDF, which cascades these errors to Trilinos all the way to a failure in the Peridigm installation. 
**NetCDF:**
Complete config log:
[NetCDF_config.log](https://github.com/peridigm/peridigm/files/4208245/NetCDF_config.log)
Collection of errors:
[NetCDF_config.log](https://github.com/peridigm/peridigm/files/4208250/NetCDF_config.log)

**Trilinos**
Installation errors:
[Trilinos_Installation_Errors.log](https://github.com/peridigm/peridigm/files/4208251/Trilinos_Installation_Errors.log)
Configuration log:
[Trilinos_Configuration.log](https://github.com/peridigm/peridigm/files/4208252/Trilinos_Configuration.log)

And at last, le grande final
**Peridigm**
Here I run into the issue of too many open files, which I solve by running it as a root user, as I am not able to increase the allowed number of files with regular permissions (Using information from StackExchange and similar sites have not helped me, like modification of the /etc/sysctl.conf file  )
The configuration works without root:
[Peridigm_Conf.log](https://github.com/peridigm/peridigm/files/4208262/Peridigm_Conf.log)
[Peridigm_Conf_Errors.log](https://github.com/peridigm/peridigm/files/4208263/Peridigm_Conf_Errors.log)

When installing as root:
[SU_Peridigm_Install.log](https://github.com/peridigm/peridigm/files/4208264/SU_Peridigm_Install.log)
[SU_Peridigm_Install_Errors.log](https://github.com/peridigm/peridigm/files/4208265/SU_Peridigm_Install_Errors.log)

Clearly, the fact that NetCDF doesn't configure with HDF5 presents a large issue, if any of you have experienced the same problem I would be glad to hear how you solved it. 
However, it seems like a bunch of other files are missing during the Trilinos installation as well, unrelated to the NetCDF issues? 

TL:DR:

- Why do my RAM and SWAP shoot through the roof when having a 80 000 point discretisation?
- What could be possible reasons for the test files missing?
- Why won't NetCDF configure using HDF5? 
- Am I breaking Trilinos in yet another way?
- Possible workarounds to the file limit problem?

**Comment by oldninja** [comment]
Hi @theodordesousa, 
I'm going to write about the second issue - the installation process. I will preface my response with the fact I'm just getting started with Peridigm too and that it also took me about a month too to get it installed. 
From your *.log files, it looks like your Peridigm installation has some issues with the MPI.
Also, it seems like you are using GCC-7.4... I tried installing all the dependencies using 7.4 because it shipped with Ubuntu 18.04 but after similar issues, I went to GCC 4.8.5 and finally was able to install it. 
If you decide to go with the route of reinstalling, this document would be of help:
[Installing](https://github.com/PeriDoX/PeriDoX/blob/master/Doc/Peridigm_Installation_Guide/Peridigm_Installation_Guide.pdf)
Best of luck!

**Comment by theodordesousa** [comment]
Thank you oldninja. So far I have tried to install as I did before, but changing to GCC 4.8.5 as per your recommendation. However I still fail to build NetCDF with HDF5 support, and haven't tried the Trilinos or Peridigm builds yet, I will update when I solve it.

**Comment by oldninja** [comment]
@theodordesousa You are very welcome. If you look at pages 24 and 38 of the linked document in my prev. comment, you will see the versions that have worked for those authors. I went for the same and worked for me in Ubuntu. My advice would be: start by trying to build to those same versions and let's see what happens next. Good luck!

**Comment by djlittl** [comment]
Regarding the memory issue for the tangent stiffness matrix, keep in mind that the number of nonzeros in the matrix is correlated with the number of bonds in the model, not the number of nodes.  So the size will increase significantly with increasing model size, and will really blow up if you hold the horizon size fixed and reduce the mesh spacing (resulting in a large number of neighbors for each node).

I suggest you try the Matrix-Free solver option, which is available for the NOX quasi-statics solver and does not construct the matrix.  See the examples in the test/regression/NOX_QS directory.

Dave

**Comment by theodordesousa** [comment]
Thank you both. 
Regarding the installation issue, I'm trying the DLR versions right now. So far the 1.5.0 combination has worked up until Trilinos, since the repository containing the old versions seems to be down. When it gets back up again, I will hopefully manage to install it.

Thank you Dave for the input regarding the memory issue. I have most definitely forgotten to reduce the horizon when increasing the the node density. 
This weekend I had the different  NOX QS solvers running, but ran into issues, where both the solver in NOX_QS_MatrixFree_NoPrec and in 
NOX_QS_MatrixFree_3x3Prec yields the following error: 
`Throw test that evaluated to true: nodeSetMap->find(deficientSetName)==nodeSetMap->end()`
` Error: The placeholder nodeset for rank deficient nodes is missing.`

In the source code, it looks like this is due to Peridigm not finding a node set called RANK_DEFICIENT_NODES. I took a guess and created this node set in the boundary conditions, but to no success. Any suggestions on how to solve this? 
I'm appending one of the yaml files which yields this error. 


```
Peridigm:

  Verbose: true
  Discretization:
      Type: "Text File"
      Input Mesh File: "domainMesh.txt"

  Materials:
      UFG:
          Material Model: "Isotropic Hardening Correspondence"
          Density: 7.850
          Young's Modulus: 210.0e9
          Poisson's Ratio: 0.31
          Yield Stress: 925.0e8
          Hardening Modulus: 4e8
          Hourglass Coefficient: 0.0
      CG:
          Material Model: "Isotropic Hardening Correspondence"
          Density: 7.850
          Young's Modulus: 210.0e9
          Poisson's Ratio: 0.31
          Hourglass Coefficient: 0.0
          Yield Stress: 185.0e8
          Hardening Modulus: 9e8
          
  Damage Models: 
    My Damage Model:
        Damage Model: "Interface Aware"
        Critical Stretch: 5.00e-02
        
  Blocks:
      My First Block:
        Block Names: "block_1"
        Material: "CG"
        Horizon: 1.15e-05
        Damage Model: "My Damage Model"
      My Second Block:
        Block Names: "block_2"
        Material: "UFG"
        Horizon: 1.15e-05
        Damage Model: "My Damage Model"
  
  Boundary Conditions:
      Node Set One: "node_set_1.txt"    
      Node Set Two: "node_set_2.txt"    
      Node Set Three: "node_set_3.txt"  
      
      # X-direction
      Prescribed Displacement Bottom In X:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "x"
        Value: "0.0"
      Prescribed Displacement Top In X:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Two"
        Coordinate: "x"
        Value: t*x*0.25/5 # 0.25 as per the curve in NiPure_2002_ASM-AtlasSScurves.pdf
      # Y-direction
      Prescribed Displacement Fix Top Rigid Body Motion In Y:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "y"
        Value: "0.0"
      Prescribed Displacement Fix Bottom Rigid Body Motion In Y:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Two"
        Coordinate: "y"
        Value: "0.0"
      Prescribed Displacement Fix Extra Rigid Body Motion In Y:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Three"
        Coordinate: "y"
        Value: "0.0"
      # Z-direction
      Prescribed Displacement Fix Bottom Rigid Body Motion In Z:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "z"
        Value: "0.0"
      Prescribed Displacement Fix Top Rigid Body Motion In Z:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Two"
        Coordinate: "z"
        Value: "0.0"
      Prescribed Displacement Fix Extra Rigid Body Motion In Z:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Three"
        Coordinate: "z"
        Value: "0.0"

  Solver:
        Verbose: true
        Initial Time: 0.0
        Final Time: 5.0
        Peridigm Preconditioner: "Block 3x3"
        NOXQuasiStatic:
            Nonlinear Solver: "Line Search Based"
            Number of Load Steps: 20
            Max Solver Iterations: 100
            Relative Tolerance: 1.0e-1
            Direction:
              Method: "Newton"
              Newton:
                Linear Solver:
                  Jacobian Operator: "Matrix-Free"
                  Preconditioner: "User Defined"
            Line Search:
              Method: "Polynomial"


  Output:
      Output File Type: "ExodusII"
      Output Format: "BINARY"
      Output Filename: "GRM_v2-h1_15e-05-cs5_00e-02-VariousNOX2"
      Output Frequency: 1
      Output Variables:
        Damage: true
        Radius: true
        Displacement: true
        Velocity: true
        Element_Id: true
        Proc_Num: true
        Force_Density: true
        Force: true
        Volume: true
        Cauchy_Stress: true
        Deformation_Gradient: true
        Number_Of_Neighbors: true
        Neighborhood_Volume: true
        Coordinates: true
        Unrotated_Rate_Of_Deformation: true
        Unrotated_Cauchy_Stress: true
        Von_Mises_Stress: true
        Equivalent_Plastic_Strain: true
        Critical_Stretch: true
        Hourglass_Force_Density: true
        Left_Stretch_Tensor: true
        Partial_Stress: true

```

**Comment by djlittl** [comment]
Looks like you’re using a text file for the mesh.  I suggest you convert this to an exodus/genesis file using the script text_to_genesis.py in the peridigm/scripts directory.  Strictly speaking, this should not be necessary.  But it will allow you to 1) visualize your input mesh in Paraview, double check that you have the node sets correct, etc., and 2) run the “decomp” utility to decompose the mesh as a pre-processing operation, as opposed to loading the mesh on the root processor and re-distributing it, which is what is done for text file meshes.

If you’re running in serial, then 2) obviously does not apply.

Generally speaking, the exodus/genesis I/O is better tested and more robust than the text file I/O.

**Comment by theodordesousa** [comment]
Perfect, thank you. Some questions regarding the exodus/genesis generation file:

- My Trilinos was compiled without shared libraries, meaning that the exodus.py module can't find any .so files, giving me an error at  `EXODUS_LIB = cdll.LoadLibrary(EXODUS_SO)`, since this tries to load ~/trilinos/lib/libexodus.so, which doesn't exist since trilinos/lib only contains .a files. 
Does this mean I have to reconfigure Trilinos with `Trilinos_BUILD_SHARED_LIBS=TRUE`?

- Is the line `sys.path.append("/Users/djlittl/Software/seacas/GCC_4.9.4_THREAD_SAFE/lib")` supposed to be to whatever thread safe gcc compiler available, or some specific seacas gcc compiler? In the latter case, there is no gcc library directory in trilinios/packages/seacas/ or its subdirectories. 

Once again, thank you for taking your free time to help me out.

**Comment by theodordesousa** [comment]
Solved the last comment I wrote now, and managed to almost install a successful installation. 
I still get four errors running the tests: 
```
187 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np1 (Failed)
188 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np4 (Failed)
189 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np1 (Failed)
190 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np4 (Failed)

```
In the log files, I can't find a clue regarding what the possible error can be, only that the file is different compared to the gold file. Note that only  isotropic hardening tests fail, and both np1 and np4 fail.

I'll append both my bash pseudo-install-script, as well as as the error logs, if someone else could extract useful information from it.

Error logs for no-flaw-test 
[IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log](https://github.com/peridigm/peridigm/files/4305681/IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log)
[IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log](https://github.com/peridigm/peridigm/files/4305682/IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw.log)


Error logs for with-flaw-test
[IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log](https://github.com/peridigm/peridigm/files/4305703/IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log)
[IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log](https://github.com/peridigm/peridigm/files/4305705/IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw.log)


The bash script I have been using to install peridigm.
[Peridigminstallation_sh_.txt](https://github.com/peridigm/peridigm/files/4305708/Peridigminstallation_sh_.txt)

**Comment by sahirbutt** [comment]
I also get the same four tests failing, with same log files as @theodordesousa showing differences in comparison with the *gold.e files.

**Comment by theodordesousa** [comment]
Sahirbutt, have you had any success?

**Comment by sahirbutt** [comment]
@theodordesousa, I still have the following tests failing

	170 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np1 (Failed)
	171 - IsotropicHardeningPlasticFullyPrescribedTension_NoFlaw_np4 (Failed)
	172 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np1 (Failed)
	173 - IsotropicHardeningPlasticFullyPrescribedTension_WithFlaw_np4 (Failed)

with *.log files saying `exodiff: Files are different` at the end.
